api/tools/src/main/java/org/pmiops/workbench/tools/institutions/User.java

```diff
@@ -0,0 +1,134 @@
+            String.format(
+                "Expected %d columns in input file. Was: %d", columnLength, headerLine.length));
+      }
+
+      return StreamSupport.stream(reader.spliterator(), false).collect(Collectors.toList());
+    }
+  }
+
+  private void checkField(String dbValue, String csvValue, String fieldName) {
+    if (!dbValue.equals(csvValue)) {
+      log.warning(
+          String.format(
+              "CSV and DB values do not match for user '%s', field '%s'. CSV = %s, DB = %s",
+              userName, fieldName, csvValue, dbValue));
+    }
+  }
+
+  private DbUser dbCheck(final UserDao userDao) {
+    final DbUser dbUser = userDao.findUserByUsername(userName);
+
+    // fatal errors: something is quite wrong and we need to recheck our assumptions!
+
+    if (dbUser == null) {
+      throw new RuntimeException(String.format("User %s was not found in the DB", userName));
+    }
+
+    // many near-matches in the imput exist: warn only, don't stop
+    // example: "Dan" instead of "Daniel"
+    // example: ""Beth,"" instead of "Beth"
+
+    checkField(dbUser.getGivenName(), firstName, "First Name");
+    checkField(dbUser.getFamilyName(), lastName, "Last Name");
+    checkField(dbUser.getContactEmail(), contactEmail, "Contact Email");
+
+    return dbUser;
+  }
+
+  void populateAffiliation(
+      final boolean dryRun,
+      final UserDao userDao,
+      final InstitutionDao institutionDao,
+      final VerifiedInstitutionalAffiliationDao affiliationDao) {
+
+    preCheck();
+
+    final DbUser dbUser = dbCheck(userDao);
+    final DbVerifiedInstitutionalAffiliation newAffiliation = toAffiliation(dbUser, institutionDao);
+    final Optional<DbVerifiedInstitutionalAffiliation> existingAffil =
+        affiliationDao.findFirstByUser(dbUser);
+
+    if (!existingAffil.isPresent()) {
+      if (!dryRun) {
+        affiliationDao.save(newAffiliation);
+      }
+
+      dryLog(
+          dryRun,
+          String.format(
+              "Saved Affiliation for '%s' with Institution '%s'",
+              newAffiliation.getUser().getUsername(),
+              newAffiliation.getInstitution().getDisplayName()));
+    } else {
+      // will always execute since we checked it above
+      existingAffil.ifPresent(
+          existingAffiliation -> {
+            if (equivalent(existingAffiliation, newAffiliation)) {
+              log.info("No action taken.  Affiliation exists: " + existingAffiliation);
+            } else {
+              throw new RuntimeException(
+                  String.format(
+                      "New affiliation differs from affiliation in DB for user '%s':\n"
+                          + "New affiliation = %s\nDB affiliation= %s",
+                      userName, newAffiliation, existingAffiliation));
+            }
+          });
+    }
+  }
+
+  // Are these two DbVerifiedInstitutionalAffiliation objects equivalent?
+  // TODO: investigate why we can't use DbVerifiedInstitutionalAffiliation.equals() here.
```

Comment:
Tearing my hair out over this one.  If I return to this, it will be here: https://github.com/all-of-us/workbench/pull/3601